library/misc/datafixerupper/src/main/java/org/quiltmc/qsl/datafixerupper/api/QuiltDataFixes.java

```diff
@@ -0,0 +1,126 @@
+
+/**
+ * Provides methods to register custom {@link DataFixer}s.
+ */
+public final class QuiltDataFixes {
+	private QuiltDataFixes() {
+		throw new RuntimeException("QuiltDataFixes only contains static declarations.");
+	}
+
+	/**
+	 * A "base" version {@code 0} schema, for use by all mods.
+	 * <p>
+	 * This schema <em>must</em> be the first one added!
+	 *
+	 * @see DataFixerBuilder#addSchema(int, BiFunction)
+	 */
+	public static final BiFunction<Integer, Schema, Schema> BASE_SCHEMA = (version, parent) -> {
+		checkArgument(version == 0, "version must be 0");
+		checkArgument(parent == null, "parent must be null");
+		return QuiltDataFixesInternals.createBaseSchema();
+	};
+
+	/**
+	 * Registers a new data fixer.
+	 *
+	 * @param modId          the mod identifier
+	 * @param currentVersion the current version of the mod's data
+	 * @param dataFixer      the data fixer
+	 */
+	public static void registerFixer(@NotNull String modId, int currentVersion, @NotNull DataFixer dataFixer) {
+		checkNotNull(modId, "modId cannot be null");
+		checkArgument(currentVersion >= 0, "currentVersion must be positive");
+		checkNotNull(dataFixer, "dataFixer cannot be null");
+
+		if (isFrozen()) {
+			throw new IllegalStateException("Can't register data fixer after registry is frozen");
+		}
+
+		QuiltDataFixesInternals.registerFixer(modId, currentVersion, dataFixer);
+	}
+
+	/**
+	 * Registers a new data fixer.
+	 *
+	 * @param mod            the mod container
+	 * @param currentVersion the current version of the mod's data
+	 * @param dataFixer      the data fixer
+	 */
+	public static void registerFixer(@NotNull ModContainer mod, int currentVersion, @NotNull DataFixer dataFixer) {
+		checkNotNull(mod, "mod cannot be null");
+
+		registerFixer(mod.metadata().id(), currentVersion, dataFixer);
+	}
+
+	/**
+	 * Gets a mod's data fixer.
+	 *
+	 * @param modId the mod identifier
+	 * @return the mod's data fixer, or empty if the mod hasn't registered one
+	 */
+	public static @NotNull Optional<DataFixer> getFixer(@NotNull String modId) {
+		checkNotNull(modId, "modId cannot be null");
+
+		QuiltDataFixesInternals.DataFixerEntry entry = QuiltDataFixesInternals.getFixerEntry(modId);
+		if (entry == null) {
+			return Optional.empty();
+		}
+		return Optional.of(entry.dataFixer());
+	}
+
+	/**
+	 * Gets a mod's data version from a {@link NbtCompound}.
+	 *
+	 * @param compound the compound
+	 * @param modId    the mod identifier
+	 * @return the mod's data version, or {@code 0} if the compound has no data for that mod
+	 */
+	public static int getModDataVersion(@NotNull NbtCompound compound, @NotNull String modId) {
+		checkNotNull(compound, "compound cannot be null");
+		checkNotNull(modId, "modId cannot be null");
+
+		return QuiltDataFixesInternals.getModDataVersion(compound, modId);
+	}
+
+	/**
+	 * Checks if the data fixer registry is frozen.
+	 *
+	 * @return {@code true} if frozen, or {@code false} otherwise.
+	 */
+	public static boolean isFrozen() {
```

Comment:
```suggestion
	@Contract(pure = true)
	public static boolean isFrozen() {
```

Would also need to annotate the internal one to make sure the contract is always respected.